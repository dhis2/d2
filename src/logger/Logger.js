import { checkType, isType } from '../lib/check'

class Logger {
    constructor(logging) {
        checkType(logging, 'object', 'console')
        this.logger = logging
    }

    canLog(type) {
        return !!(type && console && isType(this.logger[type], 'function'))
    }

    logMessage(type = 'log', ...rest) {
        if (this.canLog(type) && this.logger[type]) {
            this.logger[type](...rest)
            return true
        }
        return false
    }

    debug(...rest) {
        return this.logMessage('debug', ...rest)
    }

    error(...rest) {
        return this.logMessage('error', ...rest)
    }

    log(...rest) {
        return this.logMessage('log', ...rest)
    }

    warn(...rest) {
        return this.logMessage('warn', ...rest)
    }

    static getLogger() {
        let logger

        // TODO: This is not very clean try to figure out a better way to do this.
        try {
            // Node version
            logger = global.console
        } catch (e) {
            // Browser version fallback
            /* istanbul ignore next */
            logger = window.console
        }

        if (this.logger) {
            return this.logger
        }
        return (this.logger = new Logger(logger))
    }
}

export default Logger
