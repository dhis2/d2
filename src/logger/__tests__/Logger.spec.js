import Logger from '../Logger';

describe('Logger', () => {
    let logger;
    let consoleMock;

    beforeEach(() => {
        consoleMock = {
            log: jest.fn(),
            debug: jest.fn(),
            error: jest.fn(),
            warn: jest.fn(),
        };

        logger = new Logger(consoleMock);
    });

    it('should not be allowed to be called without new', () => {
        expect(() => Logger()).toThrowErrorMatchingSnapshot();
    });

    it('should get the correct Logger instance from the namespace', () => {
        expect(logger).toBeInstanceOf(Logger);
    });

    it('should have a log function', () => {
        expect(logger.log).toBeDefined();
        expect(logger.log).toBeInstanceOf(Function);
    });

    it('should log to the console', () => {
        logger.log('my message');

        expect(consoleMock.log).toBeCalledWith('my message');
    });

    it('should return true after successful logging', () => {
        expect(logger.log('my message')).toBe(true);
    });

    it('should not log when it does not exist', () => {
        delete consoleMock.log;

        expect(logger.log('my message')).toBe(false);
    });

    it('should not log if the method does not exist', () => {
        delete consoleMock.warn;

        expect(logger.warn('my message')).toBe(false);
    });

    it('should log a warning', () => {
        expect(logger.warn('my message')).toBe(true);
        expect(consoleMock.warn).toBeCalledWith('my message');
    });

    it('should log a debug request', () => {
        expect(logger.debug('my message')).toBe(true);
        expect(consoleMock.debug).toBeCalledWith('my message');
    });

    it('should not log when it does not exist', () => {
        delete consoleMock.debug;

        expect(logger.debug('my message')).toBe(false);
    });

    it('should log an error request', () => {
        expect(logger.error('my message')).toBe(true);
        expect(consoleMock.error).toBeCalledWith('my message');
    });

    it('should not log when error does not exist', () => {
        delete consoleMock.error;

        expect(logger.error('my message')).toBe(false);
    });

    describe('getLogger', () => {
        it('should return a logger', () => {
            expect(Logger.getLogger()).toBeInstanceOf(Logger);
        });

        it('should create a singleton and return that', () => {
            expect(Logger.getLogger()).toBe(Logger.getLogger());
        });
    });
});
