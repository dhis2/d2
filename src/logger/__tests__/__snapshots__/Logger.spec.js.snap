// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should not be allowed to be called without new 1`] = `"Class constructor Logger cannot be invoked without 'new'"`;
